applications/nucleo_io/targets/freertos.armv6m.st-stm32f091rc-nucleo-dev-board/ServoConsumer.hxx

```diff
@@ -0,0 +1,135 @@
+#ifndef _APPLICATIONS_IO_BOARD_TARGET_SERVOCONSUMER_HXX_
+#define _APPLICATIONS_IO_BOARD_TARGET_SERVOCONSUMER_HXX_
+
+#include "config.hxx"
+#include "os/MmapGpio.hxx"
+
+const unsigned servo_ticks_0 = configCPU_CLOCK_HZ * 1 / 1000;   // 48k
+const unsigned servo_ticks_180 = configCPU_CLOCK_HZ * 2 / 1000; // 96k
+
+uint8_t dummy_var[1] = {0x0};
+constexpr const MmapGpio8 DUMMY_GPIO(
+    dummy_var, /*offset=*/0, /*is_output=*/false);
+
+// GPIOBit has a <template HW> ctor that won't work if
+// supplied Gpio* is a PWMGPO; no idea why.
+class MyGPIOBit : public openlcb::BitEventInterface
+{
+public:
+    MyGPIOBit(openlcb::Node *node, openlcb::EventId event_on,
+        openlcb::EventId event_off, const Gpio *gpio)
+        : BitEventInterface(event_on, event_off)
+        , node_(node)
+        , gpio_(gpio)
+    {
+    }
+
+    openlcb::EventState get_current_state() override
+    {
+        return gpio_->is_set() ? openlcb::EventState::VALID
+                               : openlcb::EventState::INVALID;
+    }
+
+    void set_state(bool new_value) override
+    {
+        gpio_->write(new_value);
+    }
+
+    openlcb::Node *node() override
+    {
+        return node_;
+    }
+
+private:
+    openlcb::Node *node_;
+    const Gpio *gpio_;
+};
+
+// Basically a specialized ConfiguredConsumer.
+// Can't subclass ConfiguredConsumer here because ServoConsumerConfig
+// isn't a subclass of ConsumerConfig,
+class ServoConsumer : public DefaultConfigUpdateListener
+{
+public:
+    ServoConsumer(
+        openlcb::Node *node, const openlcb::ServoConsumerConfig &cfg, PWM *pwm)
+        : DefaultConfigUpdateListener()
+        , pwm_(pwm) // save for apply_config, where we actually use it.
+        , pwm_gpo_(nullptr) // not initialized until apply_config
+        , gpio_impl_(node, 0, 0, &DUMMY_GPIO)
+        , consumer_(&gpio_impl_) // don't connect consumer to PWM yet
+        , cfg_(cfg)
+    {
+    }
+
+    UpdateAction apply_configuration(
+        int fd, bool initial_load, BarrierNotifiable *done) OVERRIDE
+    {
+        AutoNotify n(done);
+
+        const openlcb::EventId cfg_event_min = cfg_.event_rotate_min().read(fd);
+        const openlcb::EventId cfg_event_max = cfg_.event_rotate_max().read(fd);
+        const uint8_t cfg_servo_min_pct = cfg_.servo_min_percent().read(fd);
```

Comment:
You want to have the ability to overdrive and underdrive servos. Some servos are able to rotate a bit more if you send them out-of-bounds signal. For this reason you want to interpret this as a signed integer, probably increase the size to 16 bit.
A good way to do this is to set minimum to be -99, maximum to be 200, and normal range be 0 to 100.

Reply:
I tested this with JMRI setting the field to "-10". OpenLCB Monitor log shows this datagram sent when I press "Write":

`
13:11:24.487: [[1a5654a5] 20 01 00 00 02 EC FF F6]  S: 02.01.12.FE.08.D3 - 05.01.01.01.18.15 Datagram: (8) 20.1.0.0.2.EC.FF.F6
`

and then the field [updates to show value "65526"](https://imgur.com/NBbIDlQ), which matches `FF F6`. When I try -11, the field shows 65525, and `FF F5` gets sent. This looks like a JMRI bug to me.